--- clapack-prefix/src/clapack/libf2c/Makefile.bak	2022-02-16 03:53:02.378868517 +0900
+++ clapack-prefix/src/clapack/libf2c/Makefile	2022-02-16 03:53:27.688868666 +0900
@@ -17,7 +17,7 @@
 SHELL = /bin/sh
 CFLAGS = -DNON_UNIX_STDIO -O3
 
-LD = wasm-ld
+LD = emcc
 RANLIB = emranlib
 AR = emar
 
@@ -31,7 +31,7 @@
 ## If your system does not have the ld command, comment out
 ## or remove both the ld and mv lines above.
 
-MISC =	f77vers.o i77vers.o main.o s_rnge.o abort_.o exit_.o getarg_.o iargc_.o\
+MISC =	f77vers.o i77vers.o s_rnge.o abort_.o exit_.o getarg_.o iargc_.o\
 	getenv_.o signal_.o s_stop.o s_paus.o system_.o cabs.o ctype.o\
 	derf_.o derfc_.o erf_.o erfc_.o sig_die.o uninit.o
 POW =	pow_ci.o pow_dd.o pow_di.o pow_hh.o pow_ii.o pow_ri.o pow_zi.o pow_zz.o
@@ -208,7 +208,7 @@
 	i77vers.c i_abs.c i_dim.c i_dnnt.c i_indx.c i_len.c i_mod.c \
 	i_nint.c i_sign.c iargc_.c iio.c ilnw.c inquire.c l_ge.c l_gt.c \
 	l_le.c l_lt.c lbitbits.c lbitshft.c libf2c.lbc libf2c.sy lio.h \
-	lread.c lwrite.c main.c makefile.sy makefile.u makefile.vc \
+	lread.c lwrite.c makefile.sy makefile.u makefile.vc \
 	makefile.wat math.hvc mkfile.plan9 open.c pow_ci.c pow_dd.c \
 	pow_di.c pow_hh.c pow_ii.c pow_qq.c pow_ri.c pow_zi.c pow_zz.c \
 	qbitbits.c qbitshft.c r_abs.c r_acos.c r_asin.c r_atan.c r_atn2.c \
